src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,325 @@
+            Debug.Assert(_instNesting != 0);
+
+            _instNesting--;
+
+            if (_firstInstArg)
+            {
+                _str.Remove(_str.Length - 1, 1);
+            }
+            else
+            {
+                Append(']');
+            }
+        }
+
+        private void OpenGenericArgument()
+        {
+            Debug.Assert(_instNesting != 0);
+
+            _nestedName = false;
+
+            if (!_firstInstArg)
+                Append(',');
+
+            _firstInstArg = false;
+
+            Append('[');
+
+            PushOpenGenericArgument();
+        }
+
+        private void CloseGenericArgument()
+        {
+            Debug.Assert(_instNesting != 0);
+
+            if (_hasAssemblySpec)
+            {
+                Append(']');
+            }
+
+            PopOpenGenericArgument();
+        }
+
+        private void AddName(string name)
+        {
+            Debug.Assert(name != null);
+
+            if (_nestedName)
+                Append('+');
+
+            _nestedName = true;
+
+            EscapeName(name);
+        }
+
+        private void AddArray(int rank)
+        {
+            Debug.Assert(rank > 0);
+
+            if (rank == 1)
+            {
+                Append("[*]");
+            }
+            else if (rank > 64)
+            {
+                // Only taken in an error path, runtime will not load arrays of more than 32 dimensions
+                _str.Append('[').Append(rank).Append(']');
+            }
+            else
+            {
+                Append('[');
+                for (int i = 1; i < rank; i++)
+                    Append(',');
+                Append(']');
+            }
+        }
+
+        private void AddAssemblySpec(string assemblySpec)
+        {
+            if (assemblySpec != null && !assemblySpec.Equals(""))
+            {
+                Append(", ");
+
+                if (_instNesting > 0)
+                {
+                    EscapeEmbeddedAssemblyName(assemblySpec);
+                }
+                else
+                {
+                    EscapeAssemblyName(assemblySpec);
+                }
+
+                _hasAssemblySpec = true;
+            }
+        }
+
+        public override string ToString()
+        {
+            Debug.Assert(_instNesting == 0);
+
+            return _str.ToString();
+        }
+
+        private static bool ContainsReservedChar(string name)
+        {
+            foreach (char c in name)
+            {
+                if (c == '\0')
+                    break;
+                if (IsTypeNameReservedChar(c))
+                    return true;
+            }
+            return false;
+        }
+
+        private static bool IsTypeNameReservedChar(char ch)
+        {
+            switch (ch)
+            {
+                case ',':
+                case '[':
+                case ']':
+                case '&':
+                case '*':
+                case '+':
+                case '\\':
```

Comment:
Would `\\` ever mean the character is being escaped already?

Reply:
No, it does escaping in below method `EscapeName(string name)`: https://github.com/dotnet/runtime/blob/d6a40b1874cac586fe28caf22843ce2324a9cae3/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeNameBuilder.cs#L174-L175